lazy loading chunks not built if bootstrap is not in main.ts file

Currently if people want to use lazy loading within Angular CLI projects and their `bootstrap` code is not placed inside of the `$PROJECT_ROOT/main.ts` file, the chunks for the lazy-loading module are not being generated.

This leads to JIT runtime errors that indicate that the lazy module couldn't be found.

```ts
core.js:1673 ERROR Error: Uncaught (in promise): Error: Cannot find module 'lazy/slider.module'
Error: Cannot find module 'lazy/slider.module'
    at $_lazy_route_resource lazy namespace object:5
```

Moving the `boostrap` call to the `main.ts` file fixes this issue (credits to @filipesilva for this finding)

--

It would be good to have some warning or note that implicates that lazy modules won't be detected if the `bootstrap` call is placed somewhere else.

Issue demonstration: https://github.com/DevVersion/lazy-load-test-cli

cc. @filipesilva

--- a/packages/angular_devkit/build_angular/test/browser/no-entry-module_spec_large.ts
+++ b/packages/angular_devkit/build_angular/test/browser/no-entry-module_spec_large.ts
@@ -6,7 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import { runTargetSpec } from '@angular-devkit/architect/testing';
+import { TestLogger, runTargetSpec } from '@angular-devkit/architect/testing';
 import { tap } from 'rxjs/operators';
 import { browserTargetSpec, host } from '../utils';
 
@@ -23,6 +23,17 @@ describe('Browser Builder no entry module', () => {
     const overrides = { baseHref: '/myUrl' };
 
     runTargetSpec(host, browserTargetSpec, overrides).pipe(
+      tap((buildEvent) => expect(buildEvent.success).toBe(true)),
+    ).toPromise().then(done, done.fail);
+  });
+
+  it('reports warning when no bootstrap code', (done) => {
+    host.replaceInFile('src/main.ts', /./g, '');
+    host.appendToFile('src/main.ts', `import './app/app.module';`);
+
+    const logger = new TestLogger('no-bootstrap');
+    runTargetSpec(host, browserTargetSpec, undefined, undefined, logger).pipe(
+      tap(() => expect(logger.includes('Lazy routes discovery is not enabled')).toBe(true)),
       tap((buildEvent) => expect(buildEvent.success).toBe(true)),
     ).toPromise().then(done, done.fail);
   });

--- a/packages/ngtools/webpack/src/angular_compiler_plugin.ts
+++ b/packages/ngtools/webpack/src/angular_compiler_plugin.ts
@@ -418,10 +418,17 @@ export class AngularCompilerPlugin {
     }
 
     // If there's still no entryModule try to resolve from mainPath.
-    if (!this._entryModule && this._mainPath) {
+    if (!this._entryModule && this._mainPath && !this._compilerOptions.enableIvy) {
       time('AngularCompilerPlugin._make.resolveEntryModuleFromMain');
       this._entryModule = resolveEntryModuleFromMain(
         this._mainPath, this._compilerHost, this._getTsProgram() as ts.Program);
+
+      if (!this.entryModule) {
+        this._warnings.push('Lazy routes discovery is not enabled. '
+          + 'Because there is neither an entryModule nor a '
+          + 'statically analyzable bootstrap code in the main file.',
+        );
+      }
       timeEnd('AngularCompilerPlugin._make.resolveEntryModuleFromMain');
     }
   }

--- a/packages/ngtools/webpack/src/entry_resolver.ts
+++ b/packages/ngtools/webpack/src/entry_resolver.ts
@@ -152,17 +152,13 @@ export function resolveEntryModuleFromMain(mainPath: string,
     .map(node => node.arguments[0] as ts.Identifier)
     .filter(node => node.kind == ts.SyntaxKind.Identifier);
 
-  if (bootstrap.length != 1) {
-    return null;
-  }
-  const bootstrapSymbolName = bootstrap[0].text;
-  const module = _symbolImportLookup(source, bootstrapSymbolName, host, program);
-  if (module) {
-    return `${module.replace(/\.ts$/, '')}#${bootstrapSymbolName}`;
+  if (bootstrap.length === 1) {
+    const bootstrapSymbolName = bootstrap[0].text;
+    const module = _symbolImportLookup(source, bootstrapSymbolName, host, program);
+    if (module) {
+      return `${module.replace(/\.ts$/, '')}#${bootstrapSymbolName}`;
+    }
   }
 
-  // shrug... something bad happened and we couldn't find the import statement.
-  throw new Error('Tried to find bootstrap code, but could not. Specify either '
-    + 'statically analyzable bootstrap code or pass in an entryModule '
-    + 'to the plugins options.');
+  return null;
 }

